### ISSUE TITLE
nx module federation increased initial build time from 90seconds to 30minutes for large angular application

### ISSUE THREAD

**Issue by deefactorial** [nrwl/nx#15864]
### Current Behavior

nx module federation promised to decrease build time for local development but the opposite was true once it was developed.

I implemented nx module federation with angular application with ~20 lazy-loaded routes and the initial build time increased from 90 seconds to 30 minutes. 

it appears that with each remote added the initial build time increases by 30 to 90 seconds. Even with just adding the largest 4 remotes it still was 20 minutes for the initial build time. If every developer must do this on the first build of the day there is no performance increase and a performance decrease, a full rebuild would be caused by a number of changes which would be a huge penalty of a 30-minute wait, like changing a script in the package.json file.

I also noticed a significant increase in angular cache sizes. 100 GB sizes of the `.angular` directory were not uncommon.

This is unacceptable and I had to revert the changes of nx module federation in favor of a full build until nx module federation addresses this issue.

my conclusion is nx module federation doesn't scale to large angular applications because each remote needs to load everything the host has. (translations, store, authentication, global services, etc), this duplication in build time on the initial build causes an exponential initial build time increase making the trade-off for incremental build time not worth it.

### Expected Behavior

I expected the build time to decrease

### Steps to Reproduce

1.
take a large angular application with 20 lazy loaded routes, and convert them to nx module federation.

### Nx Report

```shell
Node : 16.18.1
   OS   : darwin x64
   yarn : 1.22.0
   
   nx : 14.8.6
   @nrwl/angular : 14.8.6
   @nrwl/cypress : 14.8.6
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.8.6
   @nrwl/esbuild : Not Found
   @nrwl/eslint-plugin-nx : 14.8.6
   @nrwl/expo : Not Found
   @nrwl/express : Not Found
   @nrwl/jest : 14.8.6
   @nrwl/js : 14.8.6
   @nrwl/linter : 14.8.6
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : Not Found
   @nrwl/nx-cloud : 15.0.3
   @nrwl/nx-plugin : Not Found
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/rollup : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.8.6
   @nrwl/web : Not Found
   @nrwl/webpack : 14.8.6
   @nrwl/workspace : 14.8.6
   typescript : 4.7.4
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
         @ngrx/effects: 14.3.3
         @ngrx/entity: 14.3.3
         @ngrx/store: 14.3.3
         rxjs: 6.6.7
         @compodoc/compodoc: 1.1.19
         @cypress/schematic: 1.7.0
         @ngrx/schematics: 14.0.2
         @ngrx/store-devtools: 14.3.3
```

**Comment by deefactorial** [comment]
> 

Let me respond to each of your points one by one.

1. Yes, this is cold build time but there are too many cases where a complete rebuild will occur for example a simple change in the package.json file even if it doesn't affect the build in any way will trigger a complete rebuild. We do have Nx Cloud enabled on our project, but this doesn't provide any benefit. All developers have to do a complete rebuild of the whole project.

2. Yes we are using buildable libraries for all shared libraries; the only ones that are not buildable are the feature remote libraries.

3. Yes our workspace has clearly defined verticals, they are lazy-loaded routes for each individual `feature` of our application and include large chunks of unrelated code. 

4. yes when we do a build we are only doing a build of a single remote.

We are using the 3 core concepts you suggested and we still have this issue.
   1. distributed cache from nx cloud
   2. split our application into clearly defined isolated verticals
   3. use buildable libraries for shared libraries.

When I run `nx graph` the remote libraries don't depend on each other. we have refactored all cross-dependencies into shared buildable libraries.

**Comment by Coly010** [comment]
I would need a repo to investigate further. There seems to be a lot of build processes which do have some overhead, however they should be able to be parallelized and they should more often than not benefit from caching.

**Comment by deefactorial** [comment]
unfortunately, I am unable to share our private repository. At this point, I have found that `nx module federation` doesn't suit our build environment. The main benefit of being able to build remotes locally can be done without nx module federation and with a regular angular build that includes the feature remote module as the main module. This gives us the benefit of being able to build a remote locally while not increasing the initial build time of the full application. So for our use case it's better to stick to a regular build for the full application and each of the remotes. We get the benefit of faster builds for the full application and faster builds for each individual remote, without the overhead of coordinating an nx module federation build that is fragile and doesn't perform well. I think nx module federation was originally designed to be able to hot swap applications in production and we don't need this use case, the only reason I developed nx module federation was to decrease the developer build time and it did the opposite and increased the build time compared to a regular build. It appears that it nx module federation needs some time to mature to address the performance issues, so I'm going to close this for now and I will revisit this issue in a year or two when it's matured again.
